dialogue-target/src/main/java/com/palantir/dialogue/DefaultCallingThreadExecutor.java

```diff
@@ -0,0 +1,129 @@
+/*
+ * (c) Copyright 2021 Palantir Technologies Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.dialogue;
+
+import com.google.common.util.concurrent.FutureCallback;
+import com.google.common.util.concurrent.Futures;
+import com.google.common.util.concurrent.ListenableFuture;
+import com.palantir.dialogue.futures.DialogueFutures;
+import com.palantir.logsafe.Preconditions;
+import java.util.concurrent.BlockingDeque;
+import java.util.concurrent.Future;
+import java.util.concurrent.FutureTask;
+import java.util.concurrent.LinkedBlockingDeque;
+import java.util.concurrent.RejectedExecutionException;
+import java.util.concurrent.RunnableFuture;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+final class DefaultCallingThreadExecutor implements CallingThreadExecutor {
+
+    private static final Logger log = LoggerFactory.getLogger(DefaultCallingThreadExecutor.class);
+    private final long threadId = Thread.currentThread().getId();
+    private final Queue queue = new Queue();
+
+    /** Notification when main future completes. */
+    private final Runnable notifier = queue::poison;
+
+    @Override
+    public synchronized Future<?> submit(Runnable task) {
+        // if (Thread.currentThread().getId() == threadId) {
+        //     queue.checkNotPoisoned();
+        //     RunnableFuture<?> future = new FutureTask<>(task, null);
+        //     future.run();
+        //     return future;
+        // }
+
+        return queue.submit(task);
+    }
+
+    @Override
+    public void executeQueue(ListenableFuture<?> await) {
+        Preconditions.checkState(Thread.currentThread().getId() == threadId, "Executing queue on different thread");
+        Futures.addCallback(
+                await,
+                new FutureCallback<Object>() {
+                    @Override
+                    @SuppressWarnings("FutureReturnValueIgnored")
+                    public void onSuccess(Object _result) {
+                        queue.submit(notifier);
+                    }
+
+                    @Override
+                    @SuppressWarnings("FutureReturnValueIgnored")
+                    public void onFailure(Throwable _throwable) {
+                        queue.submit(notifier);
+                    }
+                },
```

Comment:
This callback can be a listener (Futures.addCallback converts the callback object into a Listener runnable which does a `get` on the completed future to call onSuccess/onFailure).
The Listener object might as well be created when DefaultCallingThreadExecutor is initialized, much like the notifier rather than per-invocation.

Reply:
These executors are created per Request, so only ever one listener will be created

Reply:
Done